### PR TITLE
Update test gen and benchmark for blockwise kr>2

### PR DESCRIPTION
**Pull request by GregoryComer**
Update test generation and benchmark logic for qb4w kernels to respect blockwise constraints - specifically that blocksize must be a multiple of 2*kr*sr and than kc must be a multiple of blocksize.

For test coverage, effectively all of the kt_lt and kt_gt tests are incompatible with this assumption. In effect, they can be all considered edge cases, but blockwise kernels don't really support any of these edge cases. As such, they aren't really applicable. The bl test, which linearly steps both bl and k by 2*kr*sr should provide sufficient coverage.

For benchmarking, I also round bl and kc up to a multiple of 2 * kr * sr. This does happen in the benchmark runner, so it is not reflected in the displayed benchmark case parameters. I arguably could put the logic in the benchmark wrapper, so this is the case. Could go either way.

### CODE REVIEW

File: tools/generate-gemm-test.py [hunk 1]
```diff
@@ -236,93 +236,52 @@ def split_ukernel_name(name):
         , test_func, isa_check)
         .loop_n(1, nr)
         .loop_m(1, mr));
-  if (k_block > 1) {
```
Comment (GregoryComer) [inline tools/generate-gemm-test.py:239]: FYI The diff is bit confusing but all that actually changed here is that I wrapped the k_lt and k_gt tests in an $if not blockwise.

**Comment by digantdesai** [comment]
> I also round bl and kc up to a multiple of 2 * kr * sr
kc should get round up already. Bl may be yes

File: bench/gemm-benchmark.h [hunk 2]
```diff
@@ -397,7 +397,7 @@ void GEMMBenchmark(benchmark::State& state,
 
   const size_t mc = state.range(0);
   const size_t nc = state.range(1);
-  const size_t kc = state.range(2);
+  const size_t kc = round_up_po2(state.range(2), 2 * kr * sr);
```
Comment (digantdesai) [inline bench/gemm-benchmark.h:400]: This is ok. And if we round this up to bl we may not have to skip. But we may introduce redundancy across tests i.e. kc + 1 * (2srkr) and kc + 2 * (2srkr) where bl = 2 * (2srkr).

File: test/gemm-microkernel-tester.cc [hunk 3]
```diff
@@ -50,32 +50,32 @@ TEST_P(GemmTest, Test) {
 
   // Loop over the `k`, `m`, and `n` values, if required.
   for (size_t k = params.loop_k_.from; k <= params.loop_k_.to;
-       k = NextPrime(k + params.loop_k_.step)) {
+       k = params.loop_k_.next(k)) {
```
Comment (digantdesai) [inline test/gemm-microkernel-tester.cc:53]: This is definitely cleaner.

File: test/gemm-microkernel-tester.h [hunk 4]
```diff
@@ -32,6 +32,25 @@
 #include <xnnpack/post-operation.h>
 #endif  // XNN_PLATFORM_JIT
 
+inline bool IsPrime(size_t n) {
```
Comment (digantdesai) [inline test/gemm-microkernel-tester.h:35]: forward declaration issue?

File: test/gemm-microkernel-tester.h [hunk 5]
```diff
@@ -511,28 +545,28 @@ struct GemmTestParams {
...
   }
-  GemmTestParams& loop_m(size_t from, size_t to, size_t step = 1) {
-    loop_m_ = LoopParams(from, to, step);
+  GemmTestParams& loop_m(size_t from, size_t to, size_t step = 1, LoopStepType step_type = LoopStepType::Linear) {
+    loop_m_ = LoopParams(from, to, step, step_type);
     return *this;
   }
-  GemmTestParams& loop_n(size_t from, size_t to, size_t step = 1) {
-    loop_n_ = LoopParams(from, to, step);
+  GemmTestParams& loop_n(size_t from, size_t to, size_t step = 1, LoopStepType step_type = LoopStepType::NextPrime) {
```
Comment (digantdesai) [inline test/gemm-microkernel-tester.h:556]: I was thinking we always default to linear and at the callsite we change enum to improve visibility. It should be same semantically just that the test code will be more explicit / readable about these jumps. IMO that's like the added benefit of your introduction of this enum. WDYT?

**Review by digantdesai — APPROVED** [review]
Minor nits. Let's push it up to Google. It should be already in digantdesai/dev. Thanks @GregoryComer.
